src/ffs/ffs.ts

```diff
@@ -1,109 +1,111 @@
-    const privTreeInstance = await privTree.resolve(privLink.cid, rootKeyName)
-    return new FFS(root, pubTreeInstance, privTreeInstance)
-  }
-
-  async put(): Promise<CID> {
-    await this.updateRoot()
-    return this.root.put()
+    return new FileSystem(root, publicTree, privateTree, key)
   }
 
-  async cid(): Promise<CID> {
-    return this.put()
+  // upgrade public IPFS folder to FileSystem
+  static async upgradePublicCID(cid: CID, keyName: string = 'filesystem-root'): Promise<FileSystem> {
+    const root = await PublicTree.empty()
+    const pubTreeInstance = await PublicTree.fromCID(cid)
+    const privTreeInstance = await PrivateTree.empty()
+    const key = await keystore.getKeyByName(keyName)
+    return new FileSystem(root, pubTreeInstance, privTreeInstance, key)
   }
 
-  async listDir(path: string): Promise<Link[] | null> {
-    const { tree, tail } = this.parsePath(path)
-    return tree.listDir(tail)
+  async ls(path: string): Promise<Link[] | null> {
+    return this.root.ls(path)
   }
 
-  async makeDir(path: string): Promise<FFS> {
-    const { tree, tail } = this.parsePath(path)
-    await tree.makeDir(tail)
-    return this.updateRoot()
+  async mkdir(path: string): Promise<CID> {
+    await this.runOnTree(path, (tree, tail) => {
+      return tree.mkdir(tail)
+    })
+    return this.sync()
   }
 
-  async addFile(path: string, content: FileContent): Promise<FFS> {
-    const { tree, tail } = this.parsePath(path)
-    await tree.addFile(tail, content)
-    return this.updateRoot()
+  async add(path: string, content: FileContent): Promise<CID> {
+    await this.runOnTree(path, (tree, tail) => {
+      return tree.add(tail, content)
+    })
+    return this.sync()
   }
 
-  async getFile(path: string): Promise<FileContent | null> {
-    const { tree, tail } = this.parsePath(path)
-    return tree.getFile(tail)
+  async cat(path: string): Promise<FileContent | null> {
+    return this.root.cat(path)
   }
 
-  async getNode(path: string): Promise<Node | null> {
-    const { tree, tail } = this.parsePath(path)
-    return tree.getNode(tail)
+  async getTree(path: string): Promise<Tree | null> {
+    return this.root.getTree(path)
   }
 
-  async updateRoot(): Promise<FFS> {
-    const pubCID = await this.pubTree.put()
-    const privCID = await this.privTree.put()
+  async sync(): Promise<CID> {
+    const pubCID = await this.publicTree.put()
+    const privCID = await this.privateTree.putEncrypted(this.key)
     const pubLink = { name: 'public', cid: pubCID }
     const privLink = { name: 'private', cid: privCID }
-    this.root.replaceLink(pubLink)
-    this.root.replaceLink(privLink)
-    return this
+    this.root = this.root
+                  .replaceLink(pubLink)
+                  .replaceLink(privLink)
+    return this.root.put()
   }
 
-  parsePath(path: string): { tree: Tree, tail: string } {
+  async runOnTree(path: string, fn: (tree: Tree, relPath: string) => Promise<Tree>): Promise<Tree> {
```

Comment:
I think this is the most confusing part of the new code. It's used in `mkdir`/`add` since we need to update the respective tree root after adding the file. Any thoughts to clean this up? The only other option I can think of is to do the same if/else/throw in both mkdir & add which could end up being pretty repetative :thinking: